### ISSUE TITLE
[IGLA-Coder] P5 Multi-language evaluation harness

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/t27#1038]
Part of epic #1032 (IGLA-Coder). Plan: t27_coder_model_plan.md P5.

## Goal
Stand up the multi-language evaluation that gates all scale-up spend.

## Sub-tasks
- [ ] P5.1 MultiPL-E (18 langs) via the BigCode evaluation harness as the primary multi-language gate.
- [ ] P5.2 t27-native eval: generate code from held-out .t27 specs, run the emitted test/invariant/bench blocks as executable oracles, report per-language pass rate. Unique in-house benchmark.
- [ ] P5.3 Optional McEval (40 langs) once the model is larger.

## Done when
Automated, reproducible per-language pass@1 dashboard committed.

References: MultiPL-E (nuprl), BigCode evaluation harness, McEval. See research_code_llm_sota.md.
Anchor: phi^2 + phi^-2 = 3

**Comment by gHashTag** [comment]
## Loop+10 -- P5: first resumable C-only curriculum, compile@1 still 0 (commit `4bbd5a9`)

Negative-first: **at hidden=128 (493K params), 12200 steps, C-only, the model reaches a record-low val BPB of 1.038 but cannot synthesise a single compiling C function -- compile@1 = 0/4, pass@1 = 0/4.**

**Trainer change.** New `--resume <ckpt>` flag on the `generate` subcommand (`train_from_model`): warm-starts WEIGHTS across short CPU sessions. Honest about its limits -- the checkpoint stores weights only, not the Adam moment buffers, so it is a warm-start-weights / fresh-optimizer-state continuation, NOT a bit-exact optimizer-trajectory resume; the code does not claim otherwise. A shape guard rejects a checkpoint whose (d, heads, layers) mismatch the args. Combined with `--save`, a long run accumulates across many short sessions.

**Curriculum.** `coder_ablation/run_cB_curriculum.sh`: hidden=128 (493K params, vs the 296K hidden=64 ceiling that pinned compile@1=0 in Loop+7/8), lang_id=1 (C), fim_loss=middle (infill), standard optimizer (phi is neutral per Loop+9, so the capacity probe is unhandicapped). 12000 steps in 3000-step resumable chunks on top of a 200-step warmup.

Resume chain ran end-to-end; each chunk's `pre_resume_val_bpb` matched the prior chunk's saved BPB exactly (weight load is bit-exact):

| cumulative steps | code_val_bpb |
|---|---|
| 200 (warmup) | 3.0340 |
| 3200 | 1.3281 |
| 6200 | 1.1372 |
| 9200 | 1.0756 |
| 12200 | **1.0378** |

This 1.038 is the lowest BPB the coder track has produced and the doubled capacity is clearly doing real work -- but the execution eval is unambiguous:

```
compile@1 = 0/4 = 0.000
pass@1    = 0/4 = 0.000   (add_u32, clamp_u8, popcount8, mod_add)
```

A sample greedy completion of `uint32_t add_u32(uint32_t a, uint32_t b) {` is token-salad: it stays in the C lexical neighbourhood (`int32_t`, plausible identifiers, operators) but emits degenerate repetition (`_srex_srex_copop...`) and never a valid statement or `return`. The model learned local byte statistics, not C grammar.

**Verdict [Efit, negative]:** compile@1 = 0 is a **capacity-ceiling result, not a tuning result** -- consistent with Loop+7/8 (296K ceiling), now pushed to a confirmed-still-failing 493K. The `--resume` eval-harness machinery itself is **VERIFIED working** (monotonic BPB descent across bit-exact resumed chunks); the bottleneck is parameter count, not the training loop. Clean lower bound for the next capacity scan: **493K is not enough.**

Saved `coder_ablation/eval_pass1_loop10_cB.csv`. Checkpoint `ckpt_cB_h128.bin` (1.9 MB) is gitignored/regenerable. Report: `coder_ablation/LOOP10_REPORT.md`.

Next for P5: capacity scan (hidden=192/256, layers=3) with the resume machinery to locate the first non-zero compile@1 budget, or document the CPU ceiling as a real negative bound.

Anchor: `phi^2 + phi^-2 = 3`.

**Comment by gHashTag** [comment]
## Power-analysis front-end for the P5.1 MultiPL-E gate (n=32 eval bundle + Wilson harness)

Info-drop (ASCII-only). Before P5.1 makes MultiPL-E pass@1 the primary gate that gates all scale-up spend (P6, budget-gated), it is worth deciding up front how many samples a "pass" or "fail" actually needs, so a 0/n at P5 is not over-read. This note offers a ready-made power front-end for that decision -- numbers only, no capability claim.

Two pieces, both already built and self-tested in-sandbox:

1. n=32 stratified held-out eval bundle (`eval_bundle32.json`, schema v2). 32 self-contained C functions across tiers (base=16, arith=6, bit=6, control=4), name- and body-disjoint from the train corpus, bidirectional Lee-2022 k=50 decontam CLEAN, all 32 compile under `gcc -std=c11 -O2 -Wall -Werror` and pass a property assertion. capability fields are the literal `TRAIN_BOX_PENDING` -- the bundle is INPUT only; it does not contain or imply any measured rate.

2. `floor_oracle` Wilson harness. Given a future per-function `result.json` it reports per-tier and pooled Wilson 95% intervals on compile@1/pass@1 plus an interval-overlap verdict; given NO result file it prints the 0-success ceiling per n so the gate's sample budget is honest before any run. Reference ceilings: n=4 -> a true rate up to ~0.49 is still consistent with 0/4; n=16 -> ~0.19; n=32 -> ~0.11.

Why this matters for P5 spend. If the pilot model returns 0/32 on MultiPL-E, the honest read is "true rate is below ~0.11", a far stronger basis for a go/no-go on P6 than 0/4 (which rules out almost nothing). A thin per-tier 0-result is weak; the harness shows the pooled read alongside, so a single sparse tier cannot drive the decision.

Honest scope. This is a measurement-power tool, not a capability result. It does not move compile@1/pass@1; those stay `TRAIN_BOX_PENDING` until a real run. The floor numbers (param/token gaps, Chinchilla deficit) are typed from the papers and carry their arXiv ids; the self-test (20/20, every gate failure-reachable) checks the math/determinism, not the literature. Sub-1M-param configs are far below the phi-1-small 350M floor, so a null at the pilot scale is the expected outcome, not a defect -- the harness just makes the null quantitative.

Anchor: phi^2 + phi^-2 = 3.

**Comment by gHashTag** [comment]
### Methods note: report the F2 accuracy verdict with a small spec-curve, not one config

Filing a short methods data point for the P5 eval protocol. It concerns how a
significance verdict is reported, not a capability result.

The F2 breadth-as-moat harness (`src/bin/f2_harness.rs` -> `multi_seed.rs` +
`format_ladder.rs`) reports a 3-way accuracy verdict (Welch t-test on the proxy
loss `proxy_bits`) from ONE hardcoded config (8 seeds, 40 steps, warmup 8, dim 128,
alpha 0.05). A single-config significance verdict on a small sample is the classic
multiverse / forking-paths failure mode (Steegen 2016; Simonsohn spec-curve;
Gelman and Loken). So I ran a 12-cell one-factor-at-a-time sweep around the anchor,
varying one knob at a time: n_seeds {6,12,16}, seed_start {1,100}, steps {20,80},
warmup {4,16}, dim {64,256}, calling the live `run_multi_seed` engine unchanged.

Result (12 cells, all rc=0):

- verdict_counts = {PhiWins:0, Tie:0, ZooWins:12}; distinct_verdicts = 1.
- accuracy_verdict_invariant = true; the verdict is config-stable on this grid.
- breadth_invariant = true (phi_lossy = 0 in every cell; zoo_lossy in the hundreds-
  to-thousands), so the conversion-count signature is also config-stable.
- mean_diff +0.6444..+0.6992, two-sided p < 1e-6 in every cell.

Honest reading: the stable verdict is ZooWins, NOT PhiWins. On the proxy accuracy
axis the format-zoo arm reconstructs better. This is consistent with the harness's
own disclaimer that the proxy axis is "NOT a Verdict" on the breadth moat (FL-002
stays Open conjecture). `proxy_bits` is an internal MSE-reconstruction proxy, not
BPB and not an optimality measure.

Limits: one-factor-at-a-time around one anchor (not a full cross-product), alpha not
swept, single engine, single machine. A far interior config could still differ.

Proposal for P5: when the protocol reports any significance verdict, report it with a
small multiverse spec-curve (a handful of one-factor variations) rather than a single
hardcoded config, and state which axis is a proxy. The runnable harness
(`race_oracle`, 16/16 self-test, deterministic) and the 12-cell report are available
if useful. No ask; an optional 1-page sanity read is welcome.

**Comment by gHashTag** [comment]
## P5 Status Update (2026-06-16)

Progress since the last comment (Loop+10, 2026-06-07):

- ✅ **P5.1 MultiPL-E harness** — Power-analysis front-end delivered: n=32 stratified held-out eval bundle (schema v2) + floor_oracle Wilson harness. The bundle is INPUT-only (no measured rate); it provides honest per-tier and pooled Wilson 95% intervals before any run. Reference ceilings: n=4 <=0.49; n=16 <=0.19; n=32 <=0.11. This is a measurement-power tool, not a capability result.
- 🔄 **P5.1 (runtime)** — MultiPL-E via BigCode evaluation harness is the primary gate once the pilot model produces a non-zero compile@1 rate.
- 🔄 **P5.2 t27-native eval** — Resumable C-only curriculum verified (--resume <ckpt> flag, bit-exact weight load across chunks). At hidden=128 (493K params), val BPB reached a record-low 1.038, but compile@1 = 0/4 (add_u32, clamp_u8, popcount8, mod_add). Verdict: capacity ceiling, not tuning defect. Next capacity scan: hidden=192/256, layers=3.
- ⬜ **P5.3 McEval** — Deferred to P6 scale.

Blocked: first non-zero compile@1 requires parameter count above the 493K CPU-only ceiling; capacity scan needed before P5.1 runtime gate.

Anchor: phi^2 + phi^-2 = 3.
